broker/src/bin/neon_broker.rs

```diff
@@ -0,0 +1,563 @@
+            }
+        }
+        Ok(())
+    }
+
+    // Add/remove subscriber according to sub_announce.
+    pub fn update_sub(&mut self, sub_announce: SubAnnounce) {
+        match sub_announce {
+            SubAnnounce::AddAll(tx) => self.subs_to_all.push(tx),
+            SubAnnounce::AddTimeline(ttid, sub_sender) => {
+                match self.subs_to_timelines.entry(ttid) {
+                    Entry::Occupied(mut o) => {
+                        let subsenders = o.get_mut();
+                        subsenders.push(sub_sender);
+                    }
+                    Entry::Vacant(v) => {
+                        v.insert(vec![sub_sender]);
+                    }
+                }
+            }
+            SubAnnounce::RemoveTimeline(ref ttid, sub_id) => {
+                remove_sub(&mut self.subs_to_timelines, ttid, sub_id);
+            }
+        }
+    }
+}
+
+impl Drop for Publisher {
+    fn drop(&mut self) {
+        self.registry.unregister_publisher(self);
+    }
+}
+
+struct NeonBrokerImpl {
+    registry: Registry,
+}
+
+#[tonic::async_trait]
+impl NeonBroker for NeonBrokerImpl {
+    async fn publish_safekeeper_info(
+        &self,
+        request: Request<tonic::Streaming<SafekeeperTimelineInfo>>,
+    ) -> Result<Response<()>, Status> {
+        let mut publisher = self.registry.register_publisher();
+
+        let mut stream = request.into_inner();
+
+        loop {
+            select! {
+                msg = stream.next() => {
+                    match msg {
+                        Some(Ok(msg)) => {publisher.send_msg(&msg)?;},
+                        Some(Err(e)) => {return Err(e);}, // grpc error from the stream
+                        None => {break;} // closed stream
+                    }
+                }
+                Some(announce) = publisher.announce_rx.next() => {
+                    publisher.update_sub(announce);
+                }
+            }
+        }
+
+        Ok(Response::new(()))
+    }
+
+    type SubscribeSafekeeperInfoStream =
+        Pin<Box<dyn Stream<Item = Result<SafekeeperTimelineInfo, Status>> + Send + 'static>>;
+
+    async fn subscribe_safekeeper_info(
+        &self,
+        request: Request<SubscribeSafekeeperInfoRequest>,
+    ) -> Result<Response<Self::SubscribeSafekeeperInfoStream>, Status> {
+        let proto_key = request
+            .into_inner()
+            .subscription_key
+            .ok_or_else(|| Status::new(Code::InvalidArgument, "missing subscription key"))?;
+        let sub_key = SubscriptionKey::from_proto_subscription_key(proto_key)?;
+        let mut subscriber = self.registry.register_subscriber(sub_key).await;
+
+        // transform rx into stream with item = Result, as method result demands
+        let output = async_stream::try_stream! {
+            while let Some(info) = subscriber.sub_rx.recv().await {
+                    yield info
+                }
```

Comment:
Small nit about the formatting since rustfmt doesn't touch these blocks.

```suggestion
                yield info
            }
```